### ISSUE TITLE
Edge does not zoom when scrolling using the 2-finger scroll gesture on a Precision Touchpad

### ISSUE THREAD

**Issue by j-hudecek** [openseadragon/openseadragon#1733]
I believe it's caused by this bug in Edge:
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7134034/

I've found this SO post suggesting to use wheel events instead.
https://stackoverflow.com/questions/42012645/mousewheel-dommousescroll-event-in-ie-edge

Is it possible to implement this workaround?

**Comment by iangilman** [comment]
Interesting! Well, it might be a matter of refining the code we use to determine which event to watch for: 

https://github.com/openseadragon/openseadragon/blob/68e9efebe699c5c47f274b572a9409d267bafaa7/src/mousetracker.js#L1000

It looks like it should be already using `wheel` on Edge... can you confirm if it's not?

**Comment by msalsbery** [comment]
Yes, we (I) need to periodically check our browser sniffing...it’s been several years.

Beyond that we shouldn’t have an issue as we detect our own set of gestures. I do need to make sure there’s not newer gestures that need to have their propagation stopped however

**Comment by j-hudecek** [comment]
For Edge navigator.appName returns Netscape but there is no "Trident" so it doesn't get recognized

**Comment by j-hudecek** [comment]
Yes, you're right, `OpenSeadragon.MouseTracker.wheelEventName` returns `wheel`.
In addition, it seems I've misread the SO post - it doesnt' say the `wheel` would fix it. 
Indeed, MS "fixed" the issue in Edge by telling people to switch to PTP which I can imagine is quite a change for OSD...
https://blogs.windows.com/msedgedev/2017/12/07/better-precision-touchpad-experience-ptp-pointer-events/#KcS1K2QcMM09oWm2.97

**Comment by iangilman** [comment]
I believe we already support pointer events in general (right, @msalsbery?) so maybe it wouldn't be too much work to add PTP support through that?

**Comment by msalsbery** [comment]
Yes, we already support pointer events and touch, so there's nothing PTP-specific to address (unless I'm missing something.

I do need to make sure we're properly detecting the availability of pointer events, however...so we're not falling back to legacy events

**Comment by iangilman** [comment]
> Yes, we already support pointer events and touch, so there's nothing PTP-specific to address (unless I'm missing something.

Well, sounds like something needs to be fixed so that the two finger scroll gesture works on the precision touch pad. I think the idea is that the fix is PTP related.

**Comment by msalsbery** [comment]
@j-hudecek Do you have an updated link to the bug in the original post?

What is the two finger scroll gesture and what do you expect to happen in an OpenSeadragon viewer when it’s used?

@iangilman I meant I didn’t find a PTP specific gesture so this issue is elsewhere (ie we need to add it perhaps?)

**Comment by iangilman** [comment]
Gotcha. Everything is interconnected!

**Comment by msalsbery** [comment]
Update: Regarding browser sniffing, I’ve confirmed Edge is picked up properly as far as features go, but we don’t have specific Edge BROWSERS enum values. My next PR adds BROWSERS.EDGE and BROWSERS.CHROMEEDGE values.

@j-hudecek Do you have an updated link to the bug in the original post?

What is the two finger scroll gesture and what do you expect to happen in an OpenSeadragon viewer when it’s used?

**Comment by msalsbery** [comment]
@j-hudecek  bump ;-)

**Comment by j-hudecek** [comment]
> What is the two finger scroll gesture

I meant this https://www.youtube.com/watch?v=BHjJu5owfAw

The issue link is dead and I can't find it anymore :(

**Comment by iangilman** [comment]
That's the same scroll gesture as Mac trackpads. I'd expect that to zoom just like the scroll wheel (assuming this is a desktop device).

**Comment by msalsbery** [comment]
@iangilman cool, I’ll add that as a new gesture

**Comment by iangilman** [comment]
You'll be able to differentiate that from a regular touchscreen two finger drag (which you should just be a pan, I believe)?

**Comment by msalsbery** [comment]
@iangilman well I haven’t put too much thought into it yet, but...

Two contacts is currently a pinch gesture if there’s a pinch handler. So I’m thinking add another two-finger gesture with associated optional handler. User can use one or the other handlers to choose which zoom method.

A more difficult approach would be to enhance pinch so if both contacts move same direction then zoom, else pinch. That’s fancy and all, but I prefer the first idea, because the pinch is flexible in that one can drag and zoom by moving separate or both fingers.

**Comment by msalsbery** [comment]
And with this new two-contact gesture, which direction is zoom in and which is zoom out??  I honestly don’t know ;)

**Comment by msalsbery** [comment]
@j-hudecek Do you have input on what you would expect with this? 

The pinch and wheel actions are pretty robust I think, so I’m not getting what exact two finger movement should do what. Whatever it is it should be relatively easy to implement, I’m just not getting the specification :)

**Comment by j-hudecek** [comment]
That's a very good question, but I'm afraid you won't like the answer:
on windows (like in the video I've posted) it scrolls down when you move the fingers down.
on mac (see https://youtu.be/Q7vK0uhqLy0?t=47) it scrolls up when you move the fingers down.
And users can configure it. To keep it simple you should treat scrolling up with 2 fingers like mousewheel up (negative deltaY)

**Comment by msalsbery** [comment]
@j-hudecek Meh, too much work...

J/K cool I can do that. I missed that you posted a video (facepalm)! Thank you for the clarification!!

Will get to it when the current MouseTracker patch lands.

**Comment by iangilman** [comment]
The thing is, on the Mac at least, this already works fine. The two finger drag on the trackpad just automatically gets turned into a scroll wheel event (taking into account the user's configuration for which direction is which), so OpenSeadragon doesn't need to know anything about it.

Looks like maybe the Windows "Precision Touchpads" don't do that directly for some reason, but they do have a specific event (not a regular touch event) you can look for. This goes into that:

https://blogs.windows.com/msedgedev/2017/12/07/better-precision-touchpad-experience-ptp-pointer-events/

If you want to support that event, that could be cool. I don't think we should be trying to interpret regular touch events to support this gesture, though, since most touch devices (such as phones and tablets) shouldn't have that gesture. The devices that do have the gesture either already work (Mac) or have a specific event (Windows precision touchpads), and in both cases you wouldn't want to reinvent the wheel (no pun intended), because you would miss some of the specifics of their implementation.

**Comment by msalsbery** [comment]
@iangilman I agree, and I was thinking the same. Thanks for the link!
